great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -1443,6 +1443,7 @@ def get_datasource(
             )
 
         if datasource_name in self._cached_datasources:
+            self._cached_datasources[datasource_name]._data_context = self
```

Comment:
The `get_datasource` and `get_asset` methods need to propagate updated `DataContext` -> `Datasource` -> `DataAsset` attributes. Otherwise you end up with missing `_data_context`s that are needed for config variable substitution and `get_validator` calls.